enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,348 @@
+  might be possible to either hardcode or autodetect this depending on the cloud
+  provider used.
+
+The **can** configurations will include the following properties:
+
+- `syncPeriod`: how often will the operator check resource usage to see if scaling
+  should be performed.
+- `gracePeriod`: how long after a successful scaling should no more scaling be
+  performed, even if the thresholds would scale the control planes again.
+- `scaleDownEnabled`: can enable or disable scaling down.
+- `scaleUpEnabled`: can enable or disable scaling up.
+- `scaleUpThresholds`: define thresholds that must be broken to trigger a scale
+  up:
+  - `cpuLoadAverage`: a percentage of averge CPU load.
+  - `memoryUsage`: a percentage of average memory usage.
+  - `timeWindow`: the time CPU and memory usage have to be above the thresholds,
+    before triggering a scale up event.
+- `scaleDownThresholds`: define thresholds that must be broken to trigger a scale
+  down:
+  - `cpuLoadAverage`: a percentage of averge CPU load.
+  - `memoryUsage`: a percentage of average memory usage.
+  - `timeWindow`: the time CPU and memory usage have to be below the thresholds,
+    before triggering a scale down event.
+
+A complete configuration in YAML format could then look like this:
+
+```yaml
+controlplaneautoscaling:
+  instanceSizePath: spec.template.spec.providerSpec.value.instanceType
+  instanceSizes:
+  - weight: 1
+    name: r4.large
+  - weight: 2
+    name: r4.xlarge
+  - weight: 4
+    name: r4.2xlarge
+  # [mandatory configuration]
+  syncPeriod: [duration as specified in golang: e.g. 5s, 15s]
+  gracePeriod: [duration as specified in golang: e.g. 5s, 15s]
+  scaledownEnabled: [true|false]
+  scaleUpThresholds:
+    cpuLoadAverage: [0-100 as percentage]
+    memoryUsage: [0-100 as percentage]
+    timeWindow: [duration as specified in golang: e.g. 5m, 15m]
+  scaleDownThresholds:
+    cpuLoadAverage: [0-100 as percentage]
+    memoryUsage: [0-100 as percentage]
+    timeWindow: [duration as specified in golang: e.g. 5m, 15m]
+```
+
+### Data gathering
+
+As specified the operator should not require access to data from operators that
+might not be installed in the cluster (e.g. `prometheus`).
+
+Instead the operator should gather the required data via the [metrics
+API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/)
+that is also used for the horizontal and vertical pod autoscalers.
+
+Data available from the metrics API includes:
+
+- CPU usage in `cores`.
+- Memory usage in a SI units.
+- The time window used to aggregate the above mentioned data.
+
+To gather data about the available CPU cores and memory available to control
+plane nodes, the operator should use the kubernetes
+[Node](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#node-v1-core)
+API.
+
+As the window returned by metrics API is likely shorter than required window for
```

Comment:
This is based on polling the metrics API at some pre defined (via user config?) Interval right?

Reply:
Exactly - that interval should be configurable via the configuration.
In this proposal the property that would control it would be `syncPeriod`.